### PR TITLE
feat(inputs.ipset): Add metric for number of entries and individual IPs

### LINKED ISSUE #16103: [inputs.ipset] add metrics for number of entries in ipsets and number of individual IPs (resolve CIDR)
### Use Case

Some automatic IP blacklisting systems (e.g. FireHOL) are automatically updating ipset contents to contain known malicious IP addresses to be blocked by firewalls. Those ipsets get updated regularly automatically to contain the latest IP blacklist data.

So, it would be nice if we could track the amount of entries of such dynamic ipsets and also see how many individual IP addresses are targeted (e.g. 10.12.5.0/8) would target 254 addresses while being just a single ipset entry.

### Expected behavior

The ipset input plugin should export metrics to contain the mentioned data.

### Actual behavior

This data is not exported.

### Additional info

I will try to provide a PR that implements that feature soon.

### PR DESCRIPTION
**Pull request by verybadsoldier**
## Summary
Some automatic blacklist systems (like e.g. FireHOL) are feeding known malicious IP addresses regularly into ipsets to be blocked by firewalls (e.g. iptables/nftables). It would be useful to be able to track the number of entries of such ipsets over time.

As entries can use CIDR notation to target a range of IPs (e.g. 10.12.4.0/8) with a single ipset entry, it is also useful to see the number of actualy IPs affected by one ipset.

## Checklist
- [x] No AI generated code was used in this PR

## Related issues
resolves #16103

### CODE REVIEW

**Review by srebhan — COMMENTED** [review]
Thanks @verybadsoldier for your contribution. I do have some comments in the code...

File: plugins/inputs/ipset/ipset.go [hunk 1]
```diff
@@ -53,9 +54,16 @@ func (i *Ipset) Gather(acc telegraf.Accumulator) error {
 		acc.AddError(e)
 	}
 
+	entryCounter := NewIpsetEntries(acc)
+
```
[the remark this answers is not in the document]
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset.go]: ok done

File: plugins/inputs/ipset/ipset_entries.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+type IpsetEntries struct {
+	acc          telegraf.Accumulator
+	initizalized bool
+	setName      string
+	numEntries   int
+	numIps       int
+}
+
+func NewIpsetEntries(acc telegraf.Accumulator) *IpsetEntries {
```
[the remark this answers is not in the document]
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset_entries.go]: ok done

File: plugins/inputs/ipset/ipset_entries.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+		ip := data[2]
+		count, err := getCountInCidr(ip)
+		if err != nil {
+			fmt.Println("Error:", err)
+			return
+		}
+		counter.numIps += count
+	}
+}
```
[the remark this answers is not in the document]
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset_entries.go]: done

File: plugins/inputs/ipset/ipset.go [hunk 4]
```diff
@@ -101,6 +109,9 @@ func (i *Ipset) Gather(acc telegraf.Accumulator) error {
 			acc.AddCounter(measurement, fields, tags)
 		}
 	}
+
+	entryCounter.commit()
```
[the remark this answers is not in the document]
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset.go]: done. But as `commit` is also used in `add_line`, `add_line` needs it also

File: plugins/inputs/ipset/ipset_entries.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+func (counter *IpsetEntries) commit() {
+	if !counter.initizalized {
+		return
+	}
+
+	fields := make(map[string]interface{}, 3)
+	fields["num_entries"] = counter.numEntries
+	fields["num_ips"] = counter.numIps
```
[the remark this answers is not in the document]
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset_entries.go]: done

File: plugins/inputs/ipset/ipset_entries_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		}
+		if acc.Metrics[index].Fields["num_entries"] != numEntries[index] {
+			t.Errorf("Expected num_entries length to be %d", numEntries[index])
+		}
+
+		if acc.Metrics[index].Fields["num_ips"] != numIps[index] {
+			t.Errorf("Expected num_ips length to be %d", numIps[index])
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset_entries_test.go]: oh hell yes :) done

File: plugins/inputs/ipset/sample.conf [hunk 7]
```diff
@@ -7,6 +7,7 @@
     ## You can avoid using sudo or root, by setting appropriate privileges for
     ## the telegraf.service systemd service.
     use_sudo = false
+    ## Add number of entries and number of individual IPs (resolve CIDR syntax) for each ipset
+    add_num_entries = false
```
[the remark this answers is not in the document]
Comment (verybadsoldier) [reply plugins/inputs/ipset/sample.conf]: done

**Comment by verybadsoldier** [comment]
Thanks for the review comments!

**Comment by verybadsoldier** [comment]
Should I resolve the conversation I think I have solved or is that up to you?

File: plugins/inputs/ipset/ipset.go [hunk 8]
```diff
@@ -133,8 +143,9 @@ func setList(timeout config.Duration, useSudo bool) (*bytes.Buffer, error) {
 func init() {
 	inputs.Add("ipset", func() telegraf.Input {
 		return &Ipset{
-			lister:  setList,
-			Timeout: defaultTimeout,
+			lister:        setList,
+			entriesParser: ipsetEntries{},
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset.go:147]: There is no need in doing this initialization as `entriesParser` should already be initialized like this.

File: plugins/inputs/ipset/ipset_entries.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+		ip := data[2]
+		count, err := getCountInCidr(ip)
+		if err != nil {
+			acc.AddError(err)
+			return
+		}
+		counter.ips += count
+	}
+}
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset_entries.go:78]: Why don't you simply return an error and add that error to the `acc` in the `Gather` caller?

File: plugins/inputs/ipset/ipset_entries.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+		counter.entries++
+
+		ip := data[2]
+		count, err := getCountInCidr(ip)
+		if err != nil {
+			acc.AddError(err)
+			return
+		}
+		counter.ips += count
+	}
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset_entries.go:77]:
```suggestion
	switch data[0] {
	case "create":
		counter.commit(acc)
		counter.initizalized = true
		counter.setName = data[1]
		counter.entries = 0
		counter.ips = 0
	case  "add":
		counter.entries++

		ip := data[2]
		count, err := getCountInCidr(ip)
		if err != nil {
			return err
		}
		counter.ips += count
	}
```

File: plugins/inputs/ipset/ipset_entries.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func (counter *ipsetEntries) commit(acc telegraf.Accumulator) {
+	if !counter.initizalized {
+		return
+	}
+
+	fields := make(map[string]interface{}, 3)
+	fields["entries"] = counter.entries
+	fields["ips"] = counter.ips
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset_entries.go:87]: Why three? I only see two fields? Anyway I think this is easier
```suggestion
	fields := map[string]interface{}{
		"entries": counter.entries,
		"ips":     counter.ips,
	}
```

File: plugins/inputs/ipset/ipset_entries.go [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+	tags := map[string]string{
+		"set": counter.setName,
+	}
+
+	acc.AddGauge(measurement, fields, tags)
+
+	// reset counter and prepare for next usage
+	counter.reset()
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset_entries.go:96]:
```suggestion
	// reset counter and prepare for next usage
	counter.initialized = false
```

**Review by srebhan — COMMENTED** [review]
Thanks for the update @verybadsoldier! A few more comments...

File: plugins/inputs/ipset/ipset_entries.go (hunk 11 above)
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset_entries.go]: Yes, was a copy/paste error.

File: plugins/inputs/ipset/ipset_entries.go (hunk 12 above)
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset_entries.go]: I thought it was nice to encapsulate everything needed for a reset in a specific function, but I changed it, no problem.

**Comment by verybadsoldier** [comment]
@srebhan 
Hey there, sorry for bothering but any updates on this? :)

File: plugins/inputs/ipset/ipset.go [hunk 13]
```diff
@@ -56,6 +58,14 @@ func (i *Ipset) Gather(acc telegraf.Accumulator) error {
 	scanner := bufio.NewScanner(out)
 	for scanner.Scan() {
 		line := scanner.Text()
+
+		if i.CountPerIPEntries {
+			err := i.entriesParser.addLine(line, acc)
+			if err != nil {
+				acc.AddError(err)
+			}
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset.go:66]: The `AddError` function already checks for the error being `nil`...
```suggestion
			acc.AddError(i.entriesParser.addLine(line, acc))
```

File: plugins/inputs/ipset/ipset_entries_test.go [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+
+	if i.addLine("add mylist 122.224.15.166 timeout 186758", acc) != nil {
+		t.Error("addLine returned an error")
+	}
+
+	if i.addLine("add mylist 47.128.40.145 timeout 431559", acc) != nil {
+		t.Error("addLine returned an error")
+	}
+
+	i.commit(acc)
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset_entries_test.go:31]: Could we please use the power of the `require` package here!? Same for the tests below...
```suggestion
	lines := []string{
		"create mylist hash:net family inet hashsize 16384 maxelem 131072 timeout 300 bucketsize 12 initval 0x4effa9ad",
		"add mylist 89.101.238.143 timeout 161558",
		"add mylist 122.224.15.166 timeout 186758",
		"add mylist 47.128.40.145 timeout 431559",
	}

	entries := ipsetEntries{}
	
	var acc testutil.Accumulator
	for _, line := range lines {
		require.NoError(t, i.addLine(line, &acc))
	}
	i.commit(acc)
```

**Review by srebhan — COMMENTED** [review]
@verybadsoldier sorry for the late reply (out sick)! I do have two suggestions, the latter should also be applied to the other tests you added...

File: plugins/inputs/ipset/ipset_entries_test.go (hunk 14 above)
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset_entries_test.go]: sure we can do that!

File: plugins/inputs/ipset/ipset.go (hunk 13 above)
Comment (verybadsoldier) [reply plugins/inputs/ipset/ipset.go]: oh, ok

**Comment by verybadsoldier** [comment]
Thanks again, latest comments have been addressed now!

File: plugins/inputs/ipset/ipset.go [hunk 15]
```diff
@@ -56,6 +58,12 @@ func (i *Ipset) Gather(acc telegraf.Accumulator) error {
 	scanner := bufio.NewScanner(out)
 	for scanner.Scan() {
 		line := scanner.Text()
+
+		if i.CountPerIPEntries {
+			err := i.entriesParser.addLine(line, acc)
+			acc.AddError(err)
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset.go:64]:
```suggestion
			acc.AddError(i.entriesParser.addLine(line, acc))
```

File: plugins/inputs/ipset/ipset_entries_test.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+	"testing"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/testutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIpsetEntries(t *testing.T) {
+	acc := new(testutil.Accumulator)
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset_entries_test.go:13]: Please use this form
```suggestion
	var acc testutil.Accumulator
```
and move this down just before the first use to limit the scope of this variable. Same for other tests.

File: plugins/inputs/ipset/ipset_entries_test.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+		"add mylist 89.101.238.143 timeout 161558",
+		"add mylist 122.224.15.166 timeout 186758",
+		"add mylist 47.128.40.145 timeout 431559",
+	}
+
+	entries := ipsetEntries{}
+	for _, line := range lines {
+		require.NoError(t, entries.addLine(line, acc))
+	}
+	entries.commit(acc)
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset_entries_test.go:26]: As the accumulator is no pointer anymore...
```suggestion
		require.NoError(t, entries.addLine(line, &acc))
	}
	entries.commit(&acc)
```

**Review by srebhan — COMMENTED** [review]
Sorry some more small things but after that we are good to go I think.

File: plugins/inputs/ipset/ipset_entries_test.go [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+			time.Unix(0, 0),
+			telegraf.Gauge,
+		),
+	}
+
+	testutil.RequireMetricsEqual(t, expected, acc.GetTelegrafMetrics(), testutil.IgnoreTime())
+}
+
+func TestIpsetEntriesCidr(t *testing.T) {
+	acc := new(testutil.Accumulator)
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset_entries_test.go:47]:
```suggestion
	var acc testutil.Accumulator
```

File: plugins/inputs/ipset/ipset_entries_test.go [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+		"add mylist1 44.128.40.145 timeout 431559",
+		"add mylist1 122.224.5.0/8 timeout 186758",
+		"add mylist1 45.128.40.145 timeout 431560",
+	}
+
+	entries := ipsetEntries{}
+	for _, line := range lines {
+		require.NoError(t, entries.addLine(line, acc))
+	}
+	entries.commit(acc)
```
Comment (srebhan) [inline plugins/inputs/ipset/ipset_entries_test.go:66]:
```suggestion
		require.NoError(t, entries.addLine(line, &acc))
	}
	entries.commit(&acc)
```

**Comment by verybadsoldier** [comment]
> Sorry some more small things but after that we are good to go I think.

Ahh I sorry, I was working on this by getting a better understanding of how interfaces work and this type receiver and pointer receiver thingie!
I see you fixed it yourself now already. Sorry, I for being late. I could have just applied your proposed changes, but I wanted to learn why they are how they are.

**Comment by srebhan** [comment]
@verybadsoldier no worries but as I'm out during holidays I wanted to get this done. :-) Thanks for your effort!

**Review by srebhan — APPROVED** [review]

**Comment by verybadsoldier** [comment]
> @verybadsoldier no worries but as I'm out during holidays I wanted to get this done. :-) Thanks for your effort!

Sure, thank you too for your time! Have a nice holiday!

**Review by DStrand1 — APPROVED** [review]

**Pull request merged**